ambry-router/src/test/java/com.github.ambry.router/AdaptiveOperationTrackerTest.java

```diff
@@ -0,0 +1,250 @@
+
+import static org.junit.Assert.*;
+
+
+/**
+ * Unit test for {@link AdaptiveOperationTracker} that tests the adaptability based on request latencies. This class
+ * only tests features not already tested in {@link OperationTrackerTest}.
+ *
+ * The status of an operation is represented as in the following format:
+ *
+ * local unsent count-local inflight count-local succeeded count-local failed count;
+ * remote unsent count-remote inflight count-remote succeeded count-remote failed count
+ *
+ * For example: 3-0-0-0; 9-0-0-0
+ */
+public class AdaptiveOperationTrackerTest {
+  private static final int REPLICA_COUNT = 6;
+  private static final int PORT = 6666;
+  private static final double QUANTILE = 0.9;
+  private static final Pair<Long, Long> LOCAL_COLO_LATENCY_RANGE = new Pair<>(0L, 100L);
+  private static final Pair<Long, Long> CROSS_COLO_LATENCY_RANGE = new Pair<>(120L, 220L);
+
+  private final List<MockDataNodeId> datanodes;
+  private final MockPartitionId mockPartition;
+  private final String localDcName;
+  private final LinkedList<ReplicaId> inflightReplicas = new LinkedList<>();
+  private final Set<ReplicaId> repetitionTracker = new HashSet<>();
+  private final Time time = new MockTime();
+  private final MetricRegistry registry = new MetricRegistry();
+  private final Histogram localColoTracker = registry.histogram("LocalColoTracker");
+  private final Histogram crossColoTracker = registry.histogram("CrossColoTracker");
+  private final Counter pastDueCounter = registry.counter("PastDueCounter");
+
+  /**
+   * Constructor that sets up state.
+   */
+  public AdaptiveOperationTrackerTest() {
+    List<Port> portList = Collections.singletonList(new Port(PORT, PortType.PLAINTEXT));
+    List<String> mountPaths = Collections.singletonList("mockMountPath");
+    datanodes = new ArrayList<>(Arrays.asList(
+        new MockDataNodeId[]{new MockDataNodeId(portList, mountPaths, "dc-0"), new MockDataNodeId(portList, mountPaths,
+            "dc-1")}));
+    localDcName = datanodes.get(0).getDatacenterName();
+    mockPartition = new MockPartitionId();
+    for (int i = 0; i < REPLICA_COUNT; i++) {
+      mockPartition.replicaIds.add(new MockReplicaId(PORT, mockPartition, datanodes.get(i % datanodes.size()), 0));
+    }
+  }
+
+  /**
+   * Tests that requests are discounted from the parallelism count once they move beyond the tolerance quantile.
+   * @throws InterruptedException
+   */
+  @Test
+  public void adaptationTest() throws InterruptedException {
+    primeTracker(localColoTracker, AdaptiveOperationTracker.MIN_DATA_POINTS_REQUIRED, LOCAL_COLO_LATENCY_RANGE);
+    primeTracker(crossColoTracker, AdaptiveOperationTracker.MIN_DATA_POINTS_REQUIRED, CROSS_COLO_LATENCY_RANGE);
+    double localColoCutoff = localColoTracker.getSnapshot().getValue(QUANTILE);
+    double crossColoCutoff = crossColoTracker.getSnapshot().getValue(QUANTILE);
+
+    OperationTracker ot = getOperationTracker(true, REPLICA_COUNT, 2);
```

Comment:
I actually want 2 here since the test case is more interesting and covers more cases with 2 rather than 1.